beacon-chain/rpc/eth/beacon/handlers_state.go

```diff
@@ -0,0 +1,343 @@
+		http2.HandleError(w, fmt.Sprintf("Could not get randao mix at index %d: %v", idx, err), http.StatusInternalServerError)
+		return
+	}
+
+	isOptimistic, err := helpers.IsOptimistic(ctx, []byte(stateId), s.OptimisticModeFetcher, s.Stater, s.ChainInfoFetcher, s.BeaconDB)
+	if err != nil {
+		http2.HandleError(w, "Could not check optimistic status: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	blockRoot, err := st.LatestBlockHeader().HashTreeRoot()
+	if err != nil {
+		http2.HandleError(w, "Could not calculate root of latest block header: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	isFinalized := s.FinalizationFetcher.IsFinalized(ctx, blockRoot)
+
+	resp := &GetRandaoResponse{
+		Data:                &Randao{Randao: hexutil.Encode(randao)},
+		ExecutionOptimistic: isOptimistic,
+		Finalized:           isFinalized,
+	}
+	http2.WriteJson(w, resp)
+}
+
+// GetSyncCommittees retrieves the sync committees for the given epoch.
+// If the epoch is not passed in, then the sync committees for the epoch of the state will be obtained.
+func (s *Server) GetSyncCommittees(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "beacon.GetSyncCommittees")
+	defer span.End()
+
+	stateId := mux.Vars(r)["state_id"]
+	if stateId == "" {
+		http2.HandleError(w, "state_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+	ok, rawEpoch, e := shared.UintFromQuery(w, r, "epoch")
+	if !ok {
+		return
+	}
+	epoch := primitives.Epoch(e)
+
+	currentSlot := s.GenesisTimeFetcher.CurrentSlot()
+	currentEpoch := slots.ToEpoch(currentSlot)
+	currentPeriodStartEpoch, err := slots.SyncCommitteePeriodStartEpoch(currentEpoch)
+	if err != nil {
+		http2.HandleError(w, fmt.Sprintf("Could not calculate start period for slot %d: %v", currentSlot, err), http.StatusInternalServerError)
+		return
+	}
+
+	requestNextCommittee := false
+	if rawEpoch != "" {
+		reqPeriodStartEpoch, err := slots.SyncCommitteePeriodStartEpoch(epoch)
+		if err != nil {
+			http2.HandleError(w, fmt.Sprintf("Could not calculate start period for epoch %d: %v", e, err), http.StatusInternalServerError)
+			return
+		}
+		if reqPeriodStartEpoch > currentPeriodStartEpoch+params.BeaconConfig().EpochsPerSyncCommitteePeriod {
+			http2.HandleError(
+				w,
+				fmt.Sprintf("Could not fetch sync committee too far in the future (requested epoch %d, current epoch %d)", e, currentEpoch),
+				http.StatusBadRequest,
+			)
+			return
+		}
+		if reqPeriodStartEpoch > currentPeriodStartEpoch {
+			requestNextCommittee = true
+			epoch = currentPeriodStartEpoch
+		}
+	}
+
+	var st state.BeaconState
+	if rawEpoch != "" {
+		st, ok = s.stateForSyncCommittee(ctx, w, &syncCommitteeStateRequest{
+			epoch:   &epoch,
+			stateId: []byte(stateId),
+		})
+		if !ok {
+			return
+		}
+	} else {
+		st, ok = s.stateForSyncCommittee(ctx, w, &syncCommitteeStateRequest{
+			epoch:   nil,
+			stateId: []byte(stateId),
+		})
+		if !ok {
+			return
+		}
+	}
```

Comment:
More of a nitpick, could this be improved like so?
```suggestion
	syncCommitteeReq := &syncCommitteeStateRequest{
		epoch:   nil,
		stateId: []byte(stateId),
	}
	if rawEpoch != "" {
		syncCommitteeReq.epoch = &epoch
	}
	st, ok = s.stateForSyncCommittee(ctx, w, syncCommitteeReq)
	if !ok {
		return
	}
```

Reply:
good idea, will change